runtime/runtime/src/verifier.rs

```diff
@@ -0,0 +1,1342 @@
+use crate::actions::{apply_rent, check_rent};
+use crate::config::{
+    safe_gas_to_balance, total_prepaid_gas, tx_cost, RuntimeConfig, TransactionCost,
+};
+use crate::{ApplyState, VerificationResult};
+use near_primitives::account::AccessKeyPermission;
+use near_primitives::errors::{
+    ActionsValidationError, InvalidAccessKeyError, InvalidTxError, ReceiptValidationError,
+    RuntimeError,
+};
+use near_primitives::receipt::{ActionReceipt, DataReceipt, Receipt, ReceiptEnum};
+use near_primitives::transaction::{
+    Action, AddKeyAction, DeleteAccountAction, DeployContractAction, FunctionCallAction,
+    SignedTransaction,
+};
+use near_primitives::utils::is_valid_account_id;
+use near_store::{get_access_key, get_account, set_access_key, set_account, TrieUpdate};
+use near_vm_logic::VMLimitConfig;
+
+/// Verifies the signed transaction on top of given state, charges the rent and transaction fees
+/// and balances, and updates the state for the used account and access keys.
+pub fn verify_and_charge_transaction(
```

Comment:
Could you explain a principle by which we decide what is a part of a Runtime implementation and what shouldn't be?

Reply:
Everything that operates within a Runtime should be part of the Runtime.
I moves this code to the `verifier` crate, since it also verifies the transaction.

It's also used by a `RuntimeAdapter` trait, so it shouldn't be part of the Runtime class.